### ISSUE TITLE
[Feature Request] `torch.autograd.cut_gradient(loss, cut_at)` — per-loss selective upstream gradient blocking with **zero extra memory / zero extra compute**

### ISSUE THREAD

**Issue by ARozental** [pytorch/pytorch#175165]
### 🚀 The feature, motivation and pitch

**Title**  
[Feature Request] `torch.autograd.cut_gradient(loss, cut_at)` — per-loss selective upstream gradient blocking with **zero extra memory / zero extra compute**

**Is this a feature request or a bug report?** Feature request

**Description**

### Problem / Motivation

Many models share a computation graph up to an intermediate tensor, then branch into multiple losses where each loss should affect *different subsets* of the network.

**Concrete example** (the exact pattern we hit constantly):

```python
state2 = backbone(input)          # expensive shared encoder / W1
hidden3 = later_shared(state2)    # W2 / decoder / shared head
loss1 = head1(hidden3, target1)   # Task 1 — should update EVERYTHING (backbone + later_shared + head1)
loss2 = head2(hidden3, target2)   # Task 2 — should update ONLY from state2 onward (NO gradients into backbone/W1)
```

Desired behavior (single forward, single backward, one optimizer):
- `W1/backbone` receives gradient **only from loss1**
- Everything from `state2` onward receives `loss1 + loss2`
- **No `retain_graph=True`** (kills peak memory on large models)
- **No second forward pass** on `later_shared` with a detached copy
- **No manual `grad` + subtract**, no temporary `requires_grad=False`, no extra compute

### Current workarounds and why they are unacceptable

1. `loss1.backward(retain_graph=True); loss2.backward()` → peak memory spike (full graph kept alive)
2. `state2_det = state2.detach(); hidden3_2 = later_shared(state2_det); loss2 = ...` → wastes compute & memory on duplicate forward
3. `torch.autograd.grad(loss2, backbone.parameters())` + manual `.sub_()` → still requires `retain_graph=True` on first backward + extra compute
4. Temporarily `requires_grad=False` on W1 → brittle, error-prone with complex modules/optimizers

When memory is the bottleneck (large backbones, long sequences, high batch size, etc.), **none of these are viable**.

### Proposed API (primary recommendation)

```python
loss2_cut = torch.autograd.cut_gradient(loss2, cut_at=state2)   # or cut_at=[state2, ...]

total_loss = loss1 + loss2_cut
total_loss.backward()          # ← normal single backward, no retain_graph, optimal memory & speed

optimizer.step()
```

**Alternative general API** (if the team prefers one call):

```python
torch.autograd.backward_with_cuts([
    (loss1, None),           # no cuts
    (loss2, state2),         # cut upstream of state2 for this loss
])
```

**Or as Tensor method** (nice ergonomics):

```python
loss2_cut = loss2.detach_upstream(state2)
```

### Minimal reproducible example (for testing)

```python
import torch
import torch.nn as nn

backbone = nn.Linear(10, 20)
later = nn.Linear(20, 30)
head1 = nn.Linear(30, 5)
head2 = nn.Linear(30, 5)

x = torch.randn(8, 10)
t1 = torch.randn(8, 5)
t2 = torch.randn(8, 5)

state2 = backbone(x)
hidden3 = later(state2)

l1 = (head1(hidden3) - t1).pow(2).mean()
l2 = (head2(hidden3) - t2).pow(2).mean()

# === with proposed feature ===
l2_cut = torch.autograd.cut_gradient(l2, cut_at=state2)
total = l1 + l2_cut
total.backward()

# Expected:
#   backbone.weight.grad comes only from l1
#   later.weight.grad comes from l1 + l2

### Why this matters (use cases)

- Multi-task / multi-head learning where auxiliary losses must not pollute the shared encoder
- Hierarchical / modular networks
- Actor-critic RL (value loss should not update early policy layers)
- Knowledge distillation / auxiliary prediction heads with controlled gradient flow
- Any time you want “one loss affects the whole model, another loss affects only downstream of X”

This is a **very common pattern** that currently has no first-class, zero-overhead solution.

### Implementation notes

The autograd engine already has the full DAG. We only need to:
- Tag gradient contributions by their originating loss root, or
- Insert a lightweight “CutGradFn” node that knows which root it belongs to and zeros upstream flow for that root only.

This would be analogous to `torch.no_grad` / `detach` but scoped to a specific loss root.

**Impact**: Huge win for memory-constrained training (large models, long contexts, high batch) and cleaner code for the entire community.

**Suggested labels**: `module: autograd`, `feature`, `triaged`
```

### Current workarounds and why they are unacceptable

1. `loss1.backward(retain_graph=True); loss2.backward()` → peak memory spike (full graph kept alive)
2. `state2_det = state2.detach(); hidden3_2 = later_shared(state2_det); loss2 = ...` → wastes compute & memory on duplicate forward
3. `torch.autograd.grad(loss2, backbone.parameters())` + manual `.sub_()` → still requires `retain_graph=True` on first backward + extra compute
4. Temporarily `requires_grad=False` on W1 → brittle, error-prone with complex modules/optimizers

When memory is the bottleneck (large backbones, long sequences, high batch size, etc.), **none of these are viable**.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93
